lib/utils/fs.go

```diff
@@ -84,27 +85,59 @@ func IsDir(path string) bool {
 
 // NormalizePath normalises path, evaluating symlinks and converting local
 // paths to absolute
-func NormalizePath(path string) (string, error) {
+func NormalizePath(path string, evaluateSymlinks bool) (string, error) {
 	s, err := filepath.Abs(path)
 	if err != nil {
 		return "", trace.ConvertSystemError(err)
 	}
-	abs, err := filepath.EvalSymlinks(s)
-	if err != nil {
-		return "", trace.ConvertSystemError(err)
+	if evaluateSymlinks {
+		s, err = filepath.EvalSymlinks(s)
+		if err != nil {
+			return "", trace.ConvertSystemError(err)
+		}
 	}
-	return abs, nil
+	return s, nil
+}
+
+// OpenFileAllowingSymlinks will open a file, if it's a symlink the file referenced by the link will be associated to
+// the returned os.File.  This will return an error if the file is not found or is a directory.
+func OpenFileAllowingSymlinks(path string) (*os.File, error) {
+	return openFile(path, true)
+}
+
+// OpenFileNoSymlinks will open a file, ensuring it's an actual file and not a directory or symlink.
+func OpenFileNoSymlinks(path string) (*os.File, error) {
+	return openFile(path, false)
 }
 
-// OpenFile opens  file and returns file handle
-func OpenFile(path string) (*os.File, error) {
-	newPath, err := NormalizePath(path)
+// openFile opens a file and returns file handle. In general symlinks should not be allowed to reduce the risk of
+// privilege escalation from Teleports elevated privileges to potentially less privileged users accidentally.
+func openFile(path string, allowSymlink bool) (*os.File, error) {
+	newPath, err := NormalizePath(path, allowSymlink)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	fi, err := os.Stat(newPath)
-	if err != nil {
-		return nil, trace.ConvertSystemError(err)
+	var fi os.FileInfo
+	if allowSymlink {
+		fi, err = os.Stat(newPath)
+		if err != nil {
+			return nil, trace.ConvertSystemError(err)
+		}
+	} else {
+		isAbs := filepath.IsAbs(newPath)
+		components := strings.Split(newPath, string(os.PathSeparator))
+		for i := 1; i <= len(components); i++ {
+			subPath := filepath.Join(components[:i]...)
+			if isAbs {
+				subPath = string(os.PathSeparator) + subPath
+			}
+			fi, err = os.Lstat(subPath)
```

Comment:
Is the Lstat skip in the first path component on purpose?

Reply:
I don't believe there is any skipping of components, can you explain further?

Reply:
The for starts with `i := 1`, so for relative paths we are not checking if the very first component is a symlink, unless I read it wrong?

Reply:
The components are selected using the sub-slice `[:i]`.  So the beginning part of the components will always be included in every iteration.  We start at `1` because the sub-slice is exclusive to the index.  This is also why the conditional is `<=`.

To me this seemed like the most obvious implementation but now I am wondering your recommendation to rebuild the path using the range of the components is easier to read.  Let me know your thoughts.

Reply:
Gotcha, I did read it wrong. Thanks for explaining.